packages/langium/src/grammar/type-system/type-collector/plain-types.ts

```diff
@@ -210,18 +210,30 @@ function plainToPropertyType(type: PlainPropertyType, union: UnionType | undefin
 }
 
 export function mergePropertyTypes(first: PlainPropertyType, second: PlainPropertyType): PlainPropertyType {
-    const flattenedFirst = flattenPlainType(first);
-    const flattenedSecond = flattenPlainType(second);
-    for (const second of flattenedSecond) {
-        if (!includesType(flattenedFirst, second)) {
-            flattenedFirst.push(second);
+    const { union: flattenedFirstUnion, array: flattenedFirstArray } = flattenPlainType(first);
+    const { union: flattenedSecondUnion, array: flattenedSecondArray } = flattenPlainType(second);
+    for (const second of flattenedSecondUnion) {
+        if (!includesType(flattenedFirstUnion, second)) {
+            flattenedFirstUnion.push(second);
```

Comment:
These two for-loops are almost the same. You could create some helper function 'unionTypes' or so.

Reply:
Plus: I prefer immutable operations, it also makes the code easier to read.